flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java

```diff
@@ -57,13 +58,30 @@ public interface SharedStateRegistry extends AutoCloseable {
      */
     StreamStateHandle registerReference(
             SharedStateRegistryKey registrationKey, StreamStateHandle state, long checkpointID);
+
+    /**
+     * Register a reference to the given checkpoint in the registry.
+     *
+     * @param checkpoint which is completed
+     */
+    void registerCompletedCheckpoint(CompletedCheckpoint checkpoint);
```

Comment:
Wouldn't it be more clear to pass the checkpoint to `unregisterUnusedStateAndCheckpoint` (and remove this method)?

That would require changing `Optional<Checkpoint>` to `List` in 
`StandaloneCompletedCheckpointStore.addCheckpointAndSubsumeOldestOne`, but that shouldn't be an issue.